crates/chain-config/src/config/codec/parquet/encode.rs

```diff
@@ -0,0 +1,466 @@
+use anyhow::anyhow;
+use std::{
+    io::Write,
+    marker::PhantomData,
+    sync::Arc,
+};
+
+use itertools::Itertools;
+use parquet::{
+    basic::Compression,
+    data_type::{
+        ByteArrayType,
+        FixedLenByteArrayType,
+        Int32Type,
+        Int64Type,
+    },
+    file::{
+        properties::WriterProperties,
+        writer::{
+            SerializedColumnWriter,
+            SerializedFileWriter,
+        },
+    },
+};
+
+use crate::{
+    config::{
+        contract_balance::ContractBalance,
+        contract_state::ContractStateConfig,
+    },
+    CoinConfig,
+    ContractConfig,
+    MessageConfig,
+};
+
+use super::schema::Schema;
+
+pub struct Encoder<W: Write + Send, T> {
+    writer: SerializedFileWriter<W>,
+    _type: PhantomData<T>,
+}
+
+impl<W: Write + Send, T: Schema> Encoder<W, T> {
+    pub fn new(writer: W, compression: Compression) -> anyhow::Result<Self> {
+        let writer = SerializedFileWriter::new(
+            writer,
+            Arc::new(T::schema()),
+            Arc::new(
+                WriterProperties::builder()
+                    .set_compression(compression)
+                    .build(),
+            ),
+        )?;
+
+        Ok(Self {
+            writer,
+            _type: PhantomData,
+        })
+    }
+}
+
+impl<W, T> Encoder<W, T>
+where
+    W: Write + Send,
+    Vec<T>: ColumnEncoder,
+{
+    pub fn write(&mut self, elements: Vec<T>) -> anyhow::Result<()> {
+        elements.encode_columns(&mut self.writer)
+    }
+
+    pub fn close(self) -> anyhow::Result<()> {
+        self.writer.close()?;
+        Ok(())
+    }
+}
+
+pub trait ColumnEncoder {
+    type ElementT: Schema;
+    fn encode_columns<W: std::io::Write + Send>(
+        &self,
+        writer: &mut SerializedFileWriter<W>,
+    ) -> anyhow::Result<()> {
+        let mut group = writer.next_row_group()?;
+
+        let num_of_columns = <Self::ElementT>::num_of_columns();
+        for index in 0..num_of_columns {
+            let mut column = group.next_column()?.ok_or_else(|| anyhow!("Error fetching column({index}). Expected there to be {num_of_columns} columns."))?;
+            self.encode_column(index, &mut column)?;
+            column.close()?;
+        }
+
+        group.close()?;
+        Ok(())
+    }
+
+    fn encode_column(
+        &self,
+        index: usize,
+        column: &mut SerializedColumnWriter<'_>,
+    ) -> anyhow::Result<()>;
+}
+
+impl ColumnEncoder for Vec<ContractConfig> {
+    type ElementT = ContractConfig;
+
+    fn encode_column(
```

Comment:
Do we really need to encode it field by field? Maybe instead, we can simply encode the whole type into a one-byte array and store it? In this case, the code will be more permanent (because we don't need to clone additionally) and flexible in terms of support.

Do we use columns somehow?

Reply:
Implemented parquet encoding with Postcard and Canonical support. Lemme know how it looks.